### ISSUE TITLE
Browser language redirect not working with Safari

### ISSUE THREAD

**Issue by Raemy92** [urbantrout/site-language-redirection#13]
For our customer we'd like to redirect french users to the french version of the website, which is working fine on Chrome, Friefox etc. but not on Safari (tested with v13).

Is there a fix or was it never the case to support Safari?

Otherwise great and simple to use extension, thanks!

**Comment by urbantrout** [comment]
Should be fixed now. 653caf15253d9fbac7002f800d8712ae315c2147

**Comment by Raemy92** [comment]
Unfortunately it's not working on our systems.

I changed the system language from German to French (Switzerland), restarted, cleared the Safari Cache and tested it but I still got to the German page first.

**Comment by urbantrout** [comment]
@Raemy92 Could you provide the request headers that Safari sends?  
In the Dev Tools right-click the request in the network panel and copy as curl.

I am particularly interested in the `Accept-Language` header.

**Comment by Raemy92** [comment]
curl 'https://www.***.ch/de/' \
-XGET \
-H 'Cookie: username=unset' \
-H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' \
-H 'Host: www.***.ch' \
-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.4 Safari/605.1.15' \
-H 'Accept-Language: fr-ch' \
-H 'Accept-Encoding: br, gzip, deflate' \
-H 'Connection: keep-alive'

**Comment by urbantrout** [comment]
Sorry, I forgot to mention, but I also need your site configuration.

**Comment by Raemy92** [comment]
What configuration exactly do you need?
If you need the language configuration, see the attached screenshot below.

![screencapture-typo3-index-php-2020-01-07-15_19_36](https://user-images.githubusercontent.com/37898303/71902081-9b945600-3161-11ea-9afc-a105e423d2a6.png)

**Comment by urbantrout** [comment]
I cannot reproduce the issue. Could you send me a link to a minimal reproducible example?

**Comment by Raemy92** [comment]
Just found out that it's not working with Firefox either when I send the Accept-Header "fr-CH". What I did is the same in Safari. I tested now both browser with "fr" only and it works. So I think there's something not working correctly with the transforming from "fr-CH" to "fr" only.

I can send you the link to the site via e-mail or pm, if you want?

**Comment by urbantrout** [comment]
I will give it another try. What version of the extension are you using?

**Comment by Raemy92** [comment]
Thanks, I'm using v2.1.3.

**Comment by urbantrout** [comment]
And what TYPO3 version?

**Comment by urbantrout** [comment]
And please provide your site configuration as yaml file. It is located under `config/sites/<sitename>/config.yaml`

**Comment by Raemy92** [comment]
9.5.13

`print_r($matchingSiteLanguage')` returns only the DE-Part:

`TYPO3\CMS\Core\Site\Entity\SiteLanguage Object ( [languageId:protected] => 0 [locale:protected] => de_DE.UTF-8 [base:protected] => TYPO3\CMS\Core\Http\Uri Object ( [scheme:protected] => https [supportedSchemes:protected] => Array ( [http] => 80 [https] => 443 ) [authority:protected] => [userInfo:protected] => [host:protected] => www.unipopfr.ch [port:protected] => [path:protected] => /de/ [query:protected] => [fragment:protected] => ) [title:protected] => Deutsch [navigationTitle:protected] => Deutsch [flagIdentifier:protected] => flags-de [twoLetterIsoCode:protected] => de [hreflang:protected] => de-DE [direction:protected] => ltr [typo3Language:protected] => de [fallbackType:protected] => strict [fallbackLanguageIds:protected] => Array ( ) [enabled:protected] => 1 [configuration:protected] => Array ( [title] => Deutsch [enabled] => 1 [languageId] => 0 [base] => /de/ [typo3Language] => de [locale] => de_DE.UTF-8 [iso-639-1] => de [navigationTitle] => Deutsch [hreflang] => de-DE [direction] => ltr [flag] => flags-de ) ) `

**Comment by urbantrout** [comment]
Quick tip. Use \TYPO3\CMS\Core\Utility\DebugUtility::debug() instead of print_r() for better output ;)

**Comment by Raemy92** [comment]
> And please provide your site configuration as yaml file. It is located under `config/sites/<sitename>/config.yaml`



[config.yaml.zip](https://github.com/urbantrout/site-language-redirection/files/4058175/config.yaml.zip)


> Quick tip. Use \TYPO3\CMS\Core\Utility\DebugUtility::debug() instead of print_r() for better output ;)

Will do the next time, thanks :)

**Comment by urbantrout** [comment]
Please update your `RedirectionMiddleware.php` file and add this after line 150:

```php
\TYPO3\CMS\Core\Utility\DebugUtility::debug($acceptLanguages, '$acceptLanguages');
\TYPO3\CMS\Core\Utility\DebugUtility::debug($siteLanguages, '$siteLanguages');
\TYPO3\CMS\Core\Utility\DebugUtility::debug($matchingSiteLanguages, '$matchingSiteLanguages');
exit;
```

Expand the output in your browser and send a screenshot please.

**Comment by Raemy92** [comment]
![image](https://user-images.githubusercontent.com/37898303/72330316-9a5dae80-36b6-11ea-865a-378edea902c8.png)

**Comment by urbantrout** [comment]
That's strange. hreflang should be `fr-FR` but it shows only `fr`in `$siteLanguages[1]`

**Comment by urbantrout** [comment]
Sorry, your site config shows, that you only set it to `fr`. The screenshot from https://github.com/urbantrout/site-language-redirection/issues/13#issuecomment-571607125 is outdated

**Comment by urbantrout** [comment]
I think I have all the information I need. I will fix the issue as soon as I can.

**Comment by Raemy92** [comment]
> Edit: I changed the "**Language Tag defined by RFC 1766 / 3066**" in the sites config from "fr-FR" to "fr" but without results.

Yeah I changed it for testing purposes and forgot to change it back. Now it's fr-FR again. Sorry for this.

**Comment by urbantrout** [comment]
> > Edit: I changed the "**Language Tag defined by RFC 1766 / 3066**" in the sites config from "fr-FR" to "fr" but without results.
> 
> Yeah I changed it for testing purposes and forgot to change it back. Now it's fr-FR again. Sorry for this.

Sorry, my bad…

**Comment by urbantrout** [comment]
I just released version 2.1.4. Could you give it a try?

**Comment by Raemy92** [comment]
Just tested it with Firefox and Safari, both with `Accept-Language: fr-CH` and it works perfectly.

Thank you very much, really appreciate it.
